domain/application/service/provider.go

```diff
@@ -449,35 +485,215 @@ func (s *ProviderService) RegisterCAASUnit(
+	// two sources in sync so instead we rely on the unique id from the
+	// provider.
+	//
+	// For CAAS this provider value lives in caas.FilesystemInfo.FilesystemId.
+	//
+	// Collect the set of provider ids and form a mapping from charm storage
+	// name.
+	providerIDs := make([]string, 0, len(providerFilesystemInfo))
+	providerIDName := make(map[string]domainstorage.Name, len(providerFilesystemInfo))
+	for _, fsInfo := range providerFilesystemInfo {
+		providerIDs = append(providerIDs, fsInfo.FilesystemId)
+		providerIDName[fsInfo.FilesystemId] = domainstorage.Name(fsInfo.StorageName)
+	}
+
+	// We fetch all existing storage instances in the model that are using one
+	// of the provider ids.
+	existingProviderStorage, err := s.st.GetStorageInstancesForProviderIDs(
+		ctx, appUUID, providerIDs,
+	)
+	if err != nil {
+		return application.RegisterUnitStorageArg{}, errors.Errorf(
+			"getting existing storage instances based on observed provider ids: %w",
+			err,
+		)
+	}
+
+	var (
+		existingUnitStorage map[domainstorage.Name][]domainstorage.StorageInstanceUUID
+		directivesToFollow  []application.StorageDirective
+	)
+
+	// Check if the unit already exists.
+	unitUUID, err := s.st.GetUnitUUIDByName(ctx, unitName)
+	if err != nil && !errors.Is(err, applicationerrors.UnitNotFound) {
+		return application.RegisterUnitStorageArg{}, errors.Errorf(
+			"checking if unit %q already exists: %w", unitName, err,
+		)
+	}
+
+	if unitUUID != "" {
+		// If the unit exists, we will get the already established storage
+		// directives including and storage that the unit currently owns.
+		directivesToFollow, existingUnitStorage, err = s.getUnitStorageInfo(
+			ctx, unitUUID,
+		)
+		if err != nil {
+			return application.RegisterUnitStorageArg{}, errors.Errorf(
+				"getting existing unit %q storage information for registration: %w",
+				unitUUID, err,
+			)
+		}
+	} else {
+		// If the unit does not exist, we will instead get and follow the
+		// storage directives of the application.
+		directivesToFollow, err = s.st.GetApplicationStorageDirectives(
+			ctx, appUUID,
+		)
+		if err != nil {
+			return application.RegisterUnitStorageArg{}, errors.Errorf(
+				"getting application %q storage directives: %w", appUUID, err,
+			)
+		}
+	}
+
+	// We need to walk through all of the storage in the model that is using one
+	// of the provider ids and make it available under the units existing
+	// storage.
+	//
+	// It is possible for storage to exist in the model that is not currently
+	// owned by a unit. This happens in CAAS models where a unit going away does
+	// not remove provisioned storage.
+	for providerId, instanceUUID := range existingProviderStorage {
+		storageName := providerIDName[providerId]
+
+		// If the storage has already been identified as belonging to the unit
+		// there is nothing more to do.
+		if slices.Contains(existingUnitStorage[storageName], instanceUUID) {
+			continue
+		}
+
+		existingUnitStorage[storageName] = append(
```

Comment:
This will panic, if the `unitUUID` is empty.

Reply:
Good catch.